### ISSUE TITLE
NameError: Missing or uninitialized constant: OpenSSL::Digest::OPENSSL_VERSION_NUMBER

### ISSUE THREAD

**Issue by byteg** [rubinius/rubinius#3563]
Using Rubinius 3.5 via RVM
Rails 4.2.4
Kununtu 15.10

bundle
--- ERROR REPORT TEMPLATE -------------------------------------------------------
- What did you do?

  I ran the command `/home/denisk/.rvm/gems/rbx-3.5@global/bin/bundle `

- What did you expect to happen?

  I expected Bundler to...

- What happened instead?

  Instead, what actually happened was...

Error details

    NameError: Missing or uninitialized constant: OpenSSL::Digest::OPENSSL_VERSION_NUMBER
      kernel/common/module.rb:654:in `const_missing'
      /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/digest.rb:19:in `__script__'
      /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/digest.rb:16:in `__script__'
      /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/digest.rb:15:in `__script__'
      kernel/common/code_loader.rb:247:in `require'
      kernel/common/kernel.rb:756:in `gem_original_require (require)'
      /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54:in `require'
      /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/rubysl/openssl.rb:20:in `__script__'
      kernel/common/code_loader.rb:247:in `require'
      kernel/common/kernel.rb:756:in `gem_original_require (require)'
      /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54:in `require'
      /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl.rb:1:in `__script__'
      kernel/common/code_loader.rb:247:in `require'
      kernel/common/kernel.rb:756:in `gem_original_require (require)'
      /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54:in `require'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/vendored_persistent.rb:4:in `__script__'
      kernel/common/code_loader.rb:247:in `require'
      kernel/common/kernel.rb:756:in `gem_original_require (require)'
      /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54:in `require'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/fetcher.rb:1:in `__script__'
      kernel/common/code_loader.rb:247:in `require'
      kernel/common/autoload.rb:56:in `resolve'
      kernel/bootstrap/rubinius.rb:132:in `synchronize'
      kernel/common/autoload.rb:52:in `resolve'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/cli/install.rb:96:in `run'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/cli.rb:172:in `install'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/cli.rb:10:in `start'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/exe/bundle:19:in `__script__'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/lib/bundler/friendly_errors.rb:7:in `with_friendly_errors'
      /home/denisk/.rvm/gems/rbx-3.5@global/gems/bundler-1.11.2/exe/bundle:17:in `__script__'
      kernel/common/kernel.rb:498:in `load'
      /home/denisk/.rvm/gems/rbx-3.5@global/bin/bundle:23:in `__script__'
      kernel/delta/code_loader.rb:66:in `load_script'
      kernel/delta/code_loader.rb:152:in `load_script'
      kernel/loader.rb:682:in `script'
      kernel/loader.rb:869:in `main'

Environment

    Bundler   1.11.2
    Rubygems  2.5.1
    Ruby      2.2.0p0 (2016-01-10 revision 0) [x86_64-linux-gnu]
    GEM_HOME  /home/denisk/.rvm/gems/rbx-3.5
    GEM_PATH  /home/denisk/.rvm/gems/rbx-3.5:/home/denisk/.rvm/gems/rbx-3.5@global
    RVM       1.26.11 (latest)
    Git       2.5.0
--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/bundler/bundler/search?q=Missing+or+uninitialized+constant%3A+OpenSSL%3A%3ADigest%3A%3AOPENSSL_VERSION_NUMBER&type=Issues

If there aren't any reports for this error yet, please create copy and paste the report template above into a new issue. Don't forget to anonymize any private data! The new issue form is located at:
https://github.com/bundler/bundler/issues/new

**Comment by brixen** [comment]
@byteg we need some more info:

1. could you please paste the result of running `ruby -v -e 'require "openssl"; p OpenSSL'`.
2. did RVM try to install a binary version of Rubinius? What was the exact command you used to install?
3. could you please paste the result of running `openssl version`.

**Comment by byteg** [comment]
@brixen thanks for reply!
1) $ruby -v -e 'require "openssl"; p OpenSSL'
rubinius 3.5 (2.2.0 97da6c2a 2016-01-10 3.4.2 JI) [x86_64-linux-gnu]
An exception occurred evaluating command line code

Could not open library /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/openssl.so - /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/openssl.so: symbol CRYPTO_memcmp, version OPENSSL_1.0.0 not defined in file libcrypto.so.1.0.0 with link time reference (LoadError::InvalidExtensionError)

Backtrace:

          Rubinius::NativeMethod.load_extension at kernel/common/native_method.rb:26
              Rubinius::CodeLoader#load_library at kernel/delta/code_loader.rb:234
                   Rubinius::CodeLoader#require at kernel/common/code_loader.rb:130
                   Rubinius::CodeLoader.require at kernel/common/code_loader.rb:241
  Kernel(Object)#gem_original_require (require) at kernel/common/kernel.rb:756
                         Kernel(Object)#require at /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54
                              Object#__script__ at /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/rubysl/openssl.rb:15
                   Rubinius::CodeLoader.require at kernel/common/code_loader.rb:247
  Kernel(Object)#gem_original_require (require) at kernel/common/kernel.rb:756
                         Kernel(Object)#require at /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54
                              Object#__script__ at /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl.rb:1
                   Rubinius::CodeLoader.require at kernel/common/code_loader.rb:247
  Kernel(Object)#gem_original_require (require) at kernel/common/kernel.rb:756
                         Kernel(Object)#require at /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54
                       { } in Object#__script__ at -e:1
    Rubinius::BlockEnvironment#call_on_instance at kernel/common/block_environment.rb:53
                  Kernel(Rubinius::Loader)#eval at kernel/common/eval.rb:159
                         Rubinius::Loader#evals at kernel/loader.rb:649
                          Rubinius::Loader#main at kernel/loader.rb:868

Caused by: Could not open library /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/openssl.so - /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl/openssl.so: symbol CRYPTO_memcmp, version OPENSSL_1.0.0 not defined in file libcrypto.so.1.0.0 with link time reference (LoadError)

       Rubinius::FFI::DynamicLibrary#initialize at kernel/platform/library.rb:370
          Rubinius::NativeMethod.load_extension at kernel/common/native_method.rb:24
              Rubinius::CodeLoader#load_library at kernel/delta/code_loader.rb:234
                   Rubinius::CodeLoader#require at kernel/common/code_loader.rb:130
                   Rubinius::CodeLoader.require at kernel/common/code_loader.rb:241
  Kernel(Object)#gem_original_require (require) at kernel/common/kernel.rb:756
                         Kernel(Object)#require at /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54
                              Object#__script__ at /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/rubysl/openssl.rb:15
                   Rubinius::CodeLoader.require at kernel/common/code_loader.rb:247
  Kernel(Object)#gem_original_require (require) at kernel/common/kernel.rb:756
                         Kernel(Object)#require at /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54
                              Object#__script__ at /home/denisk/.rvm/rubies/rbx-3.5/gems/gems/rubysl-openssl-2.8.0/lib/openssl.rb:1
                   Rubinius::CodeLoader.require at kernel/common/code_loader.rb:247
  Kernel(Object)#gem_original_require (require) at kernel/common/kernel.rb:756
                         Kernel(Object)#require at /home/denisk/.rvm/rubies/rbx-3.5/library/rubygems/core_ext/kernel_require.rb:54
                       { } in Object#__script__ at -e:1
    Rubinius::BlockEnvironment#call_on_instance at kernel/common/block_environment.rb:53
                  Kernel(Rubinius::Loader)#eval at kernel/common/eval.rb:159
                         Rubinius::Loader#evals at kernel/loader.rb:649
                          Rubinius::Loader#main at kernel/loader.rb:868




2) I used this command:
$rvm install rbx-3.5

It took binaries, here is the log:

$rvm install rbx-3.5
Searching for binary rubies, this might take some time.
Found remote file http://binaries.rubini.us/ubuntu/15.10/x86_64/rubinius-3.5.tar.bz2
Checking requirements for ubuntu.
Requirements installation successful.
rbx-3.5 - #configure
rbx-3.5 - #download
rbx-3.5 - #validate archive
rbx-3.5 - #extract
rbx-3.5 - #validate binary
rbx-3.5 - #setup
rbx-3.5 - #gemset created /home/denisk/.rvm/gems/rbx-3.5@global
rbx-3.5 - #importing gemset /home/denisk/.rvm/gemsets/global.gems...................................................
rbx-3.5 - #generating global wrappers..............
rbx-3.5 - #gemset created /home/denisk/.rvm/gems/rbx-3.5
rbx-3.5 - #importing gemsetfile /home/denisk/.rvm/gemsets/default.gems evaluated to empty gem list
rbx-3.5 - #generating default wrappers..............


3) $openssl version
OpenSSL 1.0.2d 9 Jul 2015
I tried to remove and install openssl again(it helped guys here: https://github.com/rubinius/rubinius/issues/3152). But I had no luck :(

**Comment by brixen** [comment]
@byteg so, this is what I suspected, the version that Rubinius was built with is not the version that is available when trying to load the openssl C-ext.

There are a few things you can try:

1. Build Rubinius from source without using RVM
2. Install Rubinius using [ruby-install](https://github.com/postmodern/ruby-install).
3. Try using our [Docker image](http://rubinius.com/packages/#docker)
4. Try using our Ubuntu 15.10 binary build (see http://binaries.rubinius.com/index.txt)

We can't help you with RVM issues, but we'd really like you to be able to install and run Rubinius. If you decide to build Rubinius from source and that doesn't work, please re-open the issue. Also, for long pastes, please use a Gist instead of pasting inline (the inline paste makes it very hard to follow the issue).

**Comment by brixen** [comment]
@byteg actually, option 4 won't work because of the openssl library that was installed when the binary was built. The best choice right now is to build from source for your platform. Unfortunately, dynamic linking causes terrible problems like this and there's not a simple way to fix them well. For building the Ubuntu 15.10 binaries, I'll have to set something up to use a 15.10 build image, but that's going to take work. The 2nd best option is to use the Docker image. Sorry. :(

**Comment by byteg** [comment]
@brixen sorry for my long inline pasing, i'm not using github everyday :(
Thanks for your answer, i'll try to build it from source. I've always wanted to use Rubinius in production, I hope it will be faster than MRI :)
Thanks for your work!

**Comment by brixen** [comment]
@byteg no worries at all, just pointing out using the Gist makes it easier to track the issue.

If you haven't taken a look at Docker, I highly recommend it. It's probably the best thing we have right now to give you a stable experience.

As for testing out Rubinius, let us know how it goes. You may be interested in the blog post I'm working on today about logging, metrics, and analysis.

**Comment by byteg** [comment]
@brixen i'll take a look at Docker, unfortunately I haven't used it before, thanks for your pointing.
I'll follow your blog. Now I'm using Newrelic (as many other developers I think) to make metrics in production.

**Comment by brixen** [comment]
@byteg yes, many developers use NewRelic. The two biggest drawbacks that I see for NewRelic are that 1. it instruments *your* code in a way that you do not control, and 2. it gives you a generic set of things that are graphed.

In contrast, in Rubinius the runtime itself has the instrumentation, and by default your code is not modified. However, we are also building a specific set of instrumentation instructions in Rubinius that _are guaranteed not to change the semantics of your program_. This will allow you to insert custom instrumentation tailored to what you need for your particular program. This enables you to better analyze your app's operation, something that a generic set of "memory + CPU + SQL queries" metrics doesn't provide.

Definitely check out Docker. It's super easy to get started and a great mechanism for improving a delivery pipeline.
